### ISSUE TITLE
AMQP decimal data type support missed

### ISSUE THREAD

**Issue by pinepain** [pdezwart/php-amqp#265]
Also, we need to check overall AMQP data types support and correctness.

**Comment by pinepain** [comment]
I'm not sure it could be closed as `AMQP_FIELD_KIND_DECIMAL` still not handled (see [amqp_basic_properties.c:443-L445](https://github.com/pdezwart/php-amqp/blob/df294a6bef638a50daa45f095139187e4ef6b8ac/amqp_basic_properties.c#L443-L445).

**Comment by lstrojny** [comment]
@pinepain hmmm, good point, any idea how I can trigger that code path?

**Comment by pinepain** [comment]
Less likely we can trigger decimals at this time, as we don't handle them at all: neither encode nor decode.

TL;DR

The necessity in AMQP decimal type support related mostly to other amqp tools integration. I guess python's pika can encode decimals, but for php it's hard to say how we should interpret them, they  should fit `zend_long` most time, but it's a part I'm not 100% sure about.

From [AMQP protocol spec](https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf),

```
decimal-value       = scale long-uint
scale               = OCTET                     ; number of decimal digits
```

> 1.4.1  Guidelines for Implementers:
> All numeric values are decimal unless otherwise indicated.

>4.2.5.5  Field Tables
> Decimal values are not intended to support floating point values, but rather fixed-point business values such as currency rates and amounts. They are encoded as an octet representing the number of places followed by a long signed integer. **The 'decimals' octet is not signed**.

While it's not clear for me what `All numeric values are decimal unless otherwise indicated`, the part about decimals sounds trivial - the actual value is `long-uint * 10^(-scale)` if I get it right. I'm not sure we can encode decimals at all (it's a bit unclear how to distinguish what needs to be float and what decimal without explicitly been told), but for compatibility with other tools on a protocol level we should be able to decode decimals. Another question here is what should we decode decimal into: `float`? We may loose precision by chance, so the whole idea of decimal gets ruined. In some other type e.g. from bcmath or gmp? It will lead to unnecessary deps.

Based on written above I guess as an intermediate solution we can decode decimals to floats and ignore any potential error and drop a line about that in docs. It still better than silently drop decimals as it done now.

**Comment by pinepain** [comment]
@lstrojny please, give a look to #269, at this time, I guess, it's the best we can do.

**Comment by lstrojny** [comment]
Puh, that’s a tough nut.

I think it boils down to two questions: how can we lossless map decimals in PHP. GMP doesn’t help as it only deals with integers (so yes, theoretically we could come up with some sort of encoding but that sounds crazy). That leaves us with bcmath. ext/bcmath wouldn't be too problematic as a dependency as it is shipped with PHP. But: it doesn’t actually help since bcmath only takes and returns strings but doesn’t offer e.g. an object or a resource (unlike GMP).

So what we could do then is introduce a simple `Decimal` class that does not too much:
```php
class AmqpDecimal
{
    public function __construct(string $decimal);

    public function toFloat(): float;

    public function __toString(): string;
}
```

That way we leave it up our users what they want to do with their decimals. It would also allow a user to specify that he indeed wants to send a decimal. What do you think?

**Comment by pinepain** [comment]
I haven't use decimals in PHP for ages, so I might be biased on actual use case.

`AMQPDecimal` looks great, I saw [AMQPDecimal from php-amqplib](https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Wire/AMQPDecimal.php) and I guess we can adopt it idea, but shrink down to 

```php
class AMQPDecimal
{
    protected $n;

    protected $e;

    public function __construct($n, $e)
    {
        if ($e < 0) {
            throw new AMQPException('Decimal exponent value must be unsigned!');
        }

        if ($n < 0) {
            throw new AMQPException('Decimal base value must be unsigned!');
        }

        $this->n = $n;
        $this->e = $e;
    }

    public function getE()
    {
        return $this->e;
    }

    public function getN()
    {
        return $this->n;
    }
}
```

It somehow complicates both receiving and sending a bit, it opens to those who really need it a way to represent it as they want. While returning float is a trivial, stringifying is a bit pita (unless there is already written function for that we have to `emalloc` `UINT8_MAX + 4` (sign+leading zero + dot + `UINT8_MAX` digits + `\0`) buffer and fill it one by one. It's not what I want to do in advance to adding a class. Those who need it should implement it on their side, imo.

Alternatively, we can always decode decimal as string and be happy with that, it's even simpler. Those who expect decimal there should be able to wrap it in anything they want. Sending decimal here is more tricky, or to be more precise - impossible.

The last way (but one I do like) is to handle decimals as how implemented in proposed PR - silently convert to double. I personally faces no use case for decimals in message headers and know no examples of such usage (but it clearly doesn't mean that it's not happens in the universe), so can't add anything on this more than I wrote. I'd just wait and see whether anybody even use it.

**Comment by lstrojny** [comment]
On the use case side: financial applications are typical clients of decimal encodings.

On the encoding side, I think we should have full referential transparency on `encode(decode(encode(decode(…))))` scenarios. As soon as we either decode decimals to strings we would lose that.

Although it’s more work I think we should go with a value object for decimals.

**Comment by pinepain** [comment]
Indeed. Value object sounds as the best intermediate solution at this time. Will give it a look.

**Comment by lstrojny** [comment]
@pinepain thanks so much!
